### PR TITLE
Add liquidity component

### PR DESCRIPTION
**Pull request by marshall-bean**
Screenshot 2023-04-10 at 9 42 03 AM

### CODE REVIEW

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    if (quote && address) {
+      const addLiquidityTxn = await well.addLiquidity(Object.values(amounts), quote, address);
+      const receipt = await addLiquidityTxn.wait();
+      setReceipt(receipt);
+      setQuote(null);
+      refetchBalances();
+    }
+  };
+
+  const handleInputChange = (index: number) => {
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:82]: should wrap this (and probably others) in a useCallback()

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    } else {
+      setQuote(null);
+    }
+  }, [amounts]);
+
+  const addLiquidityButtonClickHandler = async () => {
+    if (quote && address) {
+      const addLiquidityTxn = await well.addLiquidity(Object.values(amounts), quote, address);
+      const receipt = await addLiquidityTxn.wait();
+      setReceipt(receipt);
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:76]: why are you saving the receipt?

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      setQuote(null);
+      refetchBalances();
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:78]: nice

**Review by 0xalecks — CHANGES_REQUESTED** [review]

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx (hunk 2 above)
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:76]: Good point, no reason specifically more to show (temporarily) in the UI. I anticipate some future addition of toast or UI notifications and then we don't need this.

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    }
+  }, [well]);
+
+  const fetchQuote = async () => {
+    const quote = await well.addLiquidityQuote(Object.values(amounts));
+    // TODO: Temporary to show the loading effect
+    await new Promise((r) => setTimeout(r, 200));
+    setQuote(quote);
+    setLoadingQuote(false);
+  };
```
[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:58]: Giddyup

[the remark this answers is not in the document]
Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx]: Like this? https://github.com/BeanstalkFarms/Beanstalk/pull/397/commits/ff460666bbb67003675b78b747f1a857ff7bcf19

File: projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  }, [well.addLiquidity, amounts, quote, address]);
+
+  const handleInputChange = useCallback(
+    (index: number) => (a: TokenValue) => {
+      setAmounts({ ...amounts, [index]: a });
+    },
+    [amounts]
+  );
+
+  const addLiquidityButtonEnabled = address && atLeastOneAmountNonzero();
```
Comment (0xalecks) [inline projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:78]: can cache this with useMemo so it's not eval'ed every time
```suggestion
  const addLiquidityButtonEnabled = useMemo(() => address && atLeastOneAmountNonzero(), [addres, atLeastOneAmountNonzero()]);
```

Comment (marshall-bean) [reply projects/dex-ui/src/components/Liquidity/AddLiquidity.tsx:78]: Will apply in the next PR.

**Pull request merged**